appinventor/appengine/src/com/google/appinventor/client/TopToolbar.ui.xml

```diff
@@ -0,0 +1,178 @@
+      </ai:DropDownItem>
+      <hr />
+      <ai:DropDownItem name="UploadKeystore" caption="{messages.uploadKeystoreMenuItem}" visible="{hasWriteAccess}">
+        <actions:UploadKeystoreAction />
+      </ai:DropDownItem>
+      <ai:DropDownItem name="DownloadKeystore" caption="{messages.downloadKeystoreMenuItem}">
+        <actions:DownloadKeystoreAction />
+      </ai:DropDownItem>
+      <ai:DropDownItem name="DeleteKeystore" caption="{messages.deleteKeystoreMenuItem}" visible="{hasWriteAccess}">
+        <actions:DeleteKeystoreAction />
+      </ai:DropDownItem>
+    </ai:DropDownButton>
+
+    <!-- Connect Menu -->
+    <ai:DropDownButton name="ConnectTo" caption="{messages.connectTabName}"
+                       styleName="ode-TopPanelButton" ui:field="connectDropDown">
+      <ai:DropDownItem name="Wireless" caption="{messages.AICompanionMenuItem}">
+        <yaactions:WirelessAction />
+      </ai:DropDownItem>
+      <ai:DropDownItem name="Emulator" caption="{messages.emulatorMenuItem}">
+        <yaactions:EmulatorAction />
+      </ai:DropDownItem>
+      <ai:DropDownItem name="Usb" caption="{messages.usbMenuItem}">
+        <yaactions:UsbAction />
+      </ai:DropDownItem>
+      <ai:DropDownItem name="Chromebook" caption="{messages.chromebookMenuItem}">
+        <yaactions:ChromebookAction />
+      </ai:DropDownItem>
+      <hr />
+      <ai:DropDownItem name="RefreshCompanion" caption="{messages.refreshCompanionMenuItem}">
+        <yaactions:RefreshCompanionAction />
+      </ai:DropDownItem>
+      <hr />
+      <ai:DropDownItem name="Reset" caption="{messages.resetConnectionsMenuItem}">
+        <yaactions:ResetAction />
+      </ai:DropDownItem>
+      <ai:DropDownItem name="HardReset" caption="{messages.hardResetConnectionsMenuItem}">
+        <yaactions:HardResetAction />
+      </ai:DropDownItem>
+    </ai:DropDownButton>
+
+    <!-- Build Menu -->
+    <ai:DropDownButton name="Build" caption="{messages.buildTabName}"
+                       styleName="ode-TopPanelButton" ui:field="buildDropDown">
+      <ai:DropDownItem name="BuildApk" caption="{messages.showExportAndroidApk}" visible="{hasWriteAccess}">
+        <actions:BarcodeAction second="false" isAab="false" />
+      </ai:DropDownItem>
+      <ai:DropDownItem name="BuildAab" caption="{messages.showExportAndroidAab}" visible="{hasWriteAccess}">
+        <actions:BarcodeAction second="false" isAab="true" />
+      </ai:DropDownItem>
+      <hr />
+      <ai:DropDownItem name="BuildApk2" caption="{messages.showExportAndroidApk2}">
+        <actions:BarcodeAction second="true" isAab="false" />
+      </ai:DropDownItem>
+      <ai:DropDownItem name="BuildAab2" caption="{messages.showExportAndroidAab2}">
+        <actions:BarcodeAction second="true" isAab="true" />
+      </ai:DropDownItem>
+      <hr />
+      <ai:DropDownItem name="Yail" caption="{messages.generateYailMenuItem}">
+        <actions:GenerateYailAction />
+      </ai:DropDownItem>
```

Comment:
I think we either need to include `hasWriteAccess` check here or just remove the build menu entirely when in read only mode. There should be no way to modify a user's account (including overwriting a built APK) when in read only mode. Generate YAIL also modifies server side files so should not be available.